### PR TITLE
BC-4487 - error handler

### PR DESCRIPTION
**Pull request by hoeppner-dataport**
# Short Description
Improve ErrorHandling of API-Communication.

## Links to Ticket and related Pull-Requests
https://ticketsystem.dbildungscloud.de/browse/BC-4487

## Checklist before merging

- [x] QA: In addition to review, the code has been manually tested (if manual testing is possible)
- [x] PO: Any deviation from requirements was agreed with Product-Owner / ticket author / support-team
- [x] DEV: Every new component is implemented having accessibility in mind (e.g. aria-label, role property)

> Notice: Please keep this Pull-Request as a Draft (or add WIP label), until it is ready to be reviewed

### CODE REVIEW

File: src/components/data-board/BoardApi.composable.ts [hunk 1]
```diff
@@ -41,51 +43,29 @@ export const useBoardApi = () => {
 		return response.data;
 	};
 
-	const updateCardHeightCall = async (
-		id: string,
-		height: number
-	): Promise<number> => {
-		const response = await cardsApi.cardControllerUpdateCardHeight(id, {
+	const updateCardHeightCall = async (id: string, height: number) => {
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/data-board/BoardApi.composable.ts:46]: Not for this ticket. Would like to get that ticket closed :)

File: src/components/util-board/BoardNotifier.composable.ts [hunk 2]
```diff
@@ -1,12 +1,19 @@
...
+// WIP: move to @/types/
+export type ErrorType =
+	| "notCreated"
+	| "notLoaded"
+	| "notUpdated"
+	| "notDeleted";
 
-export const useBoardNotifier = () => {
-	const i18n = injectStrict(I18N_KEY);
+export type BoardObjectType =
```
Comment (NFriedo) [inline src/components/util-board/BoardNotifier.composable.ts:10]: Do we still need the `BoardObjectType` here?

File: src/components/util-board/BoardNotifier.composable.ts [hunk 3]
```diff
@@ -1,12 +1,19 @@
-import { HttpStatusCode } from "@/store/types/http-status-code.enum";
-import { I18N_KEY, NOTIFIER_MODULE_KEY, injectStrict } from "@/utils/inject";
+import { NOTIFIER_MODULE_KEY, injectStrict } from "@/utils/inject";
 
-type ErrorTypes = "create" | "read" | "update" | "delete";
-type BoardObjectTypes = "board" | "boardColumn" | "boardCard" | "boardElement";
+// WIP: move to @/types/
```
Comment (NFriedo) [inline src/components/util-board/BoardNotifier.composable.ts:3]: Will this WIP be solved in this ticket? When not should we create a follow up ticket with the the leftovers so we don't forget them?

File: src/components/data-board/BoardApi.composable.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (NFriedo) [reply src/components/data-board/BoardApi.composable.ts:46]: Okay maybe we can create a follow up ticket with the leftovers :)

File: src/components/data-board/BoardState.composable.unit.ts [hunk 4]
```diff
@@ -74,249 +91,252 @@ describe("BoardState.composable", () => {
...
+				expect(mockedErrorHandlerCalls.handleError).toHaveBeenCalled();
+			});
 		});
 	});
 
 	describe("createColumn", () => {
-		it("should not call createCardCall when board value is undefined", async () => {
-			const { createColumn, board } = setup();
-			board.value = undefined;
+		describe("when everything works as expected", () => {
```
Comment (NFriedo) [inline src/components/data-board/BoardState.composable.unit.ts:149]: I am not sure about this describe block. It isn't really specific. Do we have a better description for it or do we really need it? 🤔

File: src/components/util-board/BoardNotifier.composable.ts (hunk 3 above)
Comment (hoeppner-dataport) [reply src/components/util-board/BoardNotifier.composable.ts]: Moved the types to a better place: into the ErrorHandler.composable
And removed the comment.

File: src/components/data-board/CardRequestPool.composable.ts [hunk 5]
```diff
@@ -66,11 +67,14 @@ const useCardRequestPool = () => {
 					requestEntry.resolve(card);
 				}
 			}
+			const cardIds: string[] = cards.map((card) => card.id);
+			const failed = cardRequests.filter(({ id }) => !cardIds.includes(id));
+			failed.forEach((requestEntry) => requestEntry.reject(new Error()));
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/data-board/CardRequestPool.composable.ts:72]: The requestEntry carries a resolve and reject of the original promise (from the place where fetchCard was called. Line 72 calls for all failedRequests the corresponding reject-callback. The error handling should then happen around the fetchCard call...

File: src/components/error-handling/ErrorHandler.composable.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
+import { useI18n } from "@/composables/i18n.composable";
+import { ApiResponseError, ApiValidationError } from "@/store/types/commons";
+import { mapAxiosErrorToResponseError } from "@/utils/api";
+import { BoardObjectType, ErrorType, useBoardNotifier } from "@util-board";
+
+// WIP: move type
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/error-handling/ErrorHandler.composable.ts]: fixed

File: src/components/feature-board/card/CardHost.vue [hunk 7]
```diff
@@ -116,7 +116,7 @@ export default defineComponent({
 		height: { type: Number, required: true },
 		cardId: { type: String, required: true },
 	},
-	emits: ["move:card-keyboard", "delete:card"],
+	emits: ["move:card-keyboard", "delete:card", "reload:board"],
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board/card/CardHost.vue:119]: I think so, as the emit of CardHost.vue is used to emit this type of event.

File: src/components/data-board/BoardState.composable.unit.ts (hunk 4 above)
Comment (hoeppner-dataport) [reply src/components/data-board/BoardState.composable.unit.ts:149]: fixed

File: src/components/feature-board/board/BoardColumn.vue [hunk 8]
```diff
@@ -156,6 +157,10 @@ export default defineComponent({
 			emit("move:column-keyboard", event);
 		};
 
+		const onReloadBoard = () => {
+			emit("reload:board");
+		};
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board/board/BoardColumn.vue:162]: fixed

File: src/components/feature-board/board/Board.vue [hunk 9]
```diff
@@ -175,6 +179,10 @@ export default defineComponent({
 			}
 		};
 
+		const onReloadBoard = async () => {
+			await reloadBoard();
+		};
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board/board/Board.vue:184]: added

File: src/components/feature-board/board/BoardColumn.vue (hunk 8 above)
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board/board/BoardColumn.vue:162]: added

File: src/components/error-handling/ErrorHandler.composable.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+			const text = generateErrorText(errorType, boardObjectType);
+			showCustomNotifier(text, status, timeout);
+		};
+	};
+
+	const defaultErrorMap: ErrorMap = {
+		// WIP: define additional errorCode defaults ? should that all be applicationErrors
+		404: notifyWithTemplate("notLoaded"),
+		500: notifyWithTemplate("notLoaded"),
+	}; // WIP: define better default notifications than "notLoaded"
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/error-handling/ErrorHandler.composable.ts:60]: will remain for the moment

File: src/components/error-handling/ErrorHandler.composable.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+		timeout?: number
+	): ApiErrorHandler => {
+		return () => {
+			const text = generateErrorText(errorType, boardObjectType);
+			showCustomNotifier(text, status, timeout);
+		};
+	};
+
+	const defaultErrorMap: ErrorMap = {
+		// WIP: define additional errorCode defaults ? should that all be applicationErrors
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/error-handling/ErrorHandler.composable.ts:57]: will remain for the moment

**Review by NFriedo — APPROVED** [review]

**Pull request merged**